File: TalentSystemDefaults/TalentSystemDefaults/DataAccess/DataObjects/TableObjects/tbl_payment_type_bu.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using TalentSystemDefaults.DataEntities;
namespace TalentSystemDefaults
{
	namespace DataAccess.DataObjects
	{
		public class tbl_payment_type_bu : DBObjectBase
		{
			#region Class Level Fields
			private const string TBL_PAYMENT_TYPE_BU = "tbl_payment_type_bu";
			#endregion
			#region Shared Fields
			static private bool _VariableKey1Active = true;
			public static bool VariableKey1Active
			{
				get
				{
					return _VariableKey1Active;
				}
				set
				{
					_VariableKey1Active = value;
				}
			}
			static private bool _VariableKey2Active = false;
			public static bool VariableKey2Active
			{
				get
				{
					return _VariableKey2Active;
				}
				set
				{
					_VariableKey2Active = value;
				}
			}
			static private bool _VariableKey3Active = false;
			public static bool VariableKey3Active
			{
				get
				{
					return _VariableKey3Active;
				}
				set
				{
					_VariableKey3Active = value;
				}
			}
			static private bool _VariableKey4Active = false;
			public static bool VariableKey4Active
			{
				get
				{
					return _VariableKey4Active;
				}
				set
				{
					_VariableKey4Active = value;
				}
			}
			static private bool _BaseDefinition = true;
			public static bool BaseDefinition
			{
				get
				{
					return _BaseDefinition;
				}
				set
				{
					_BaseDefinition = value;
				}
			}
			#endregion
			#region Private Properties
			private string paymentTypeCode = string.Empty;
			#endregion
			#region Constructors
			/// <summary>
			/// Constructor
			/// </summary>
			/// <param name="settings"></param>
			/// <remarks></remarks>
			public tbl_payment_type_bu(ref DESettings settings)
				: base(settings)
			{
				businessUnit = settings.BusinessUnit;
				if (settings.VariableKey1 != string.Empty)
				{
					paymentTypeCode = settings.VariableKey1;
				}
			}
			#endregion
			#region Public Methods
			public override void RetrieveCurrentValues(System.Collections.Generic.List<ConfigurationItem> configs)
			{
				string currentValue = "";
				var currentValues = RetrieveCurrentValues(businessUnit, configs);
				foreach (ConfigurationItem config in configs)
				{
					if (currentValues.ContainsKey(config.DefaultName))
					{
						currentValue = currentValues[config.DefaultName];
						config.CurrentValue = currentValue;
						config.UpdatedValue = currentValue;
					}
				}
			}
			public Dictionary<string, string> RetrieveDefaultPaymentTypes(string businessUnit, string[] paymentTypeCodes)
			{
				DataTable outputDataTable = new DataTable();
				TalentDataAccess talentSqlAccessDetail = new TalentDataAccess();
				int length = paymentTypeCodes.Length;
				System.Collections.Generic.List<string> @params = new System.Collections.Generic.List<string>();
				for (int i = 0; i <= length - 1; i++)
				{
					@params.Add("@PAYMENT_TYPE_CODE" + i.ToString());
				}
				string commandText = string.Format("SELECT PAYMENT_TYPE_CODE, DEFAULT_PAYMENT_TYPE FROM tbl_payment_type_bu WHERE business_unit = @BUSINESS_UNIT AND payment_type_code IN ({0})", string.Join(",", @params.ToArray()));
				ErrorObj err = new ErrorObj();
				try
				{
					//Construct The Call
					talentSqlAccessDetail.Settings = settings;
					talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteDataSet;
					talentSqlAccessDetail.CommandElements.CommandParameter.Clear();
					talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BUSINESS_UNIT", businessUnit));
					for (int i = 0; i <= length - 1; i++)
					{
						talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter(@params[i], paymentTypeCodes[i]));
					}
					talentSqlAccessDetail.CommandElements.CommandText = commandText;
					err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005);
					if ((!(err.HasError)) && (!(talentSqlAccessDetail.ResultDataSet == null)))
					{
						outputDataTable = talentSqlAccessDetail.ResultDataSet.Tables[0];
					}
				}
				catch (Exception)
				{
					throw;
				}
				finally
				{
					talentSqlAccessDetail = null;
				}
				Dictionary<string, string> results = new Dictionary<string, string>();
				foreach (DataRow row in outputDataTable.Rows)
				{
					string key = row["PAYMENT_TYPE_CODE"].ToString();
					string value = row["DEFAULT_PAYMENT_TYPE"].ToString();
					if (!results.ContainsKey(key))
					{
						results.Add(key, value);
					}
				}
				return results;
			}
			public override int InsertCurrentValue(System.Collections.Generic.List<ConfigurationItem> configs, SqlTransaction givenTransaction, ref string key)
			{
				return InsertData(businessUnit, paymentTypeCode, configs, givenTransaction);
			}
			public override int UpdateCurrentValue(System.Collections.Generic.List<ConfigurationItem> configs, SqlTransaction givenTransaction = null)
			{
				return UpdateData(businessUnit, paymentTypeCode, configs, givenTransaction);
			}
			public override int DeleteCurrentValues(SqlTransaction givenTransaction = null)
			{
				return DeleteData(businessUnit, settings.VariableKey1, givenTransaction);
			}
			public override List<ConfigurationEntity> RetrieveAllValues(string businessUnit, string[] defaultKeys, string[] variableKeys, string displayName = "")
			{
				DataTable outputDataTable = new DataTable();
				TalentDataAccess talentSqlAccessDetail = new TalentDataAccess();
				string commandText = string.Format("SELECT * FROM tbl_payment_type_bu WHERE business_unit=@BUSINESS_UNIT AND payment_type_code=@PAYMENT_TYPE_CODE");
				try
				{
					//Construct The Call
					talentSqlAccessDetail.Settings = settings;
					talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteDataSet;
					talentSqlAccessDetail.CommandElements.CommandParameter.Clear();
					talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BUSINESS_UNIT", businessUnit));
					talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@PAYMENT_TYPE_CODE", variableKeys[0]));
					ErrorObj err = new ErrorObj();
					talentSqlAccessDetail.CommandElements.CommandText = commandText;
					err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005);
					if ((!(err.HasError)) && (!(talentSqlAccessDetail.ResultDataSet == null)))
					{
						outputDataTable = talentSqlAccessDetail.ResultDataSet.Tables[0];
					}
				}
				catch (Exception)
				{
					throw;
				}
				finally
				{
					talentSqlAccessDetail = null;
				}
				return GetConfigurationData(outputDataTable, defaultKeys, variableKeys);
			}
			#endregion
			#region Private Methods
			/// <summary>
			/// Returns configurations as list of ConfigurationEntity read from dataTable passed
			/// </summary>
			/// <param name="dt"></param>
			/// <param name="defaultKeys"></param>
			/// <param name="variableKeys"></param>
			/// <returns></returns>
			/// <remarks></remarks>
			private List<ConfigurationEntity> GetConfigurationData(DataTable dt, string[] defaultKeys, string[] variableKeys)
			{
				var results = new System.Collections.Generic.List<ConfigurationEntity>();
				foreach (DataRow row in dt.Rows)
				{
					foreach (System.Data.DataColumn column in dt.Columns)
					{
						string columnName = column.ColumnName;
						string dispName = columnName;
						var defaultName = columnName;
						string defaultValue = Utilities.CheckForDBNull_String(row[columnName]);
						results.Add(new ConfigurationEntity(TBL_PAYMENT_TYPE_BU, string.Empty, defaultKeys, variableKeys, dispName, defaultName, defaultValue, string.Empty, string.Empty));
					}
				}
				return results;
			}
			/// <summary>
			/// Returns set of configurations read from datatable passed
			/// </summary>
			/// <param name="dt"></param>
			/// <param name="configs"></param>
			/// <returns></returns>
			/// <remarks></remarks>
			private Dictionary<string, string> GetCurrentValues(DataTable dt, System.Collections.Generic.List<ConfigurationItem> configs)
			{
				var results = new Dictionary<string, string>();
				if (dt.Rows.Count > 0)
				{
					DataRow row = dt.Rows[0];
					string key = string.Empty;
					string value = string.Empty;
					foreach (System.Data.DataColumn column in dt.Columns)
					{
						key = column.ColumnName;
						value = Utilities.CheckForDBNull_String(row[key]);
						results.Add(key, value);
					}
				}
				return results;
			}
			/// <summary>
			/// Returns current configurations
			/// </summary>
			/// <param name="businessUnit"></param>
			/// <param name="configs"></param>
			/// <returns></returns>
			/// <remarks></remarks>
			private Dictionary<string, string> RetrieveCurrentValues(string businessUnit, System.Collections.Generic.List<ConfigurationItem> configs)
			{
				DataTable outputDataTable = new DataTable();
				TalentDataAccess talentSqlAccessDetail = new TalentDataAccess();
				System.Collections.Generic.List<string> codes = new System.Collections.Generic.List<string>();
				string commandText = "SELECT * FROM tbl_payment_type_bu WHERE payment_type_code = @PAYMENT_TYPE_CODE AND business_unit = @BUSINESS_UNIT";
				ErrorObj err = new ErrorObj();
				try
				{
					//Construct The Call
					talentSqlAccessDetail.Settings = settings;
					talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteDataSet;
					talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@PAYMENT_TYPE_CODE", paymentTypeCode));
					talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BUSINESS_UNIT", businessUnit));
					talentSqlAccessDetail.CommandElements.CommandText = commandText;
					err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005);
					if ((!(err.HasError)) && (!(talentSqlAccessDetail.ResultDataSet == null)))
					{
						outputDataTable = talentSqlAccessDetail.ResultDataSet.Tables[0];
					}
				}
				catch (Exception)
				{
					throw;
				}
				finally
				{
					talentSqlAccessDetail = null;
				}
				return GetCurrentValues(outputDataTable, configs);
			}
			/// <summary>
			/// Updates configurations
			/// </summary>
			/// <param name="businessUnit"></param>
			/// <param name="paymentCode"></param>
			/// <param name="configs"></param>
			/// <param name="givenTransaction"></param>
			/// <returns></returns>
			/// <remarks></remarks>
			private int UpdateData(string businessUnit, string paymentCode, System.Collections.Generic.List<ConfigurationItem> configs, SqlTransaction givenTransaction)
			{
				string[] setValue = new string[configs.Count - 1 + 1];
				string name = "";
				for (int i = 0; i <= configs.Count - 1; i++)
				{
					name = configs[i].DefaultName;
					setValue[i] = string.Format("{0}  =  @VAR_{0}", name);
				}
				string values = string.Join(" , ", setValue);
				var commandText = string.Format("UPDATE tbl_payment_type_bu SET {0} WHERE business_unit = @BUSINESS_UNIT AND payment_type_code = @PAYMENT_TYPE_CODE", values);
				int affectedRows = 0;
				TalentDataAccess talentSqlAccessDetail = new TalentDataAccess();
				ErrorObj err = new ErrorObj();
				//Construct The Call
				talentSqlAccessDetail.Settings = settings;
				talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteNonQuery;
				talentSqlAccessDetail.CommandElements.CommandText = commandText;
				for (int i = 0; i <= configs.Count - 1; i++)
				{
					talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter(string.Format("@VAR_{0}", configs[i].DefaultName), configs[i].UpdatedValue));
				}
				talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BUSINESS_UNIT", businessUnit));
				talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@PAYMENT_TYPE_CODE", paymentCode));
				//Execute
				if (givenTransaction == null)
				{
					err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005);
				}
				else
				{
					err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005, givenTransaction);
				}
				if ((!(err.HasError)) && (!(talentSqlAccessDetail.ResultDataSet == null)))
				{
					affectedRows = System.Convert.ToInt32(talentSqlAccessDetail.ResultDataSet.Tables[0].Rows[0][0]);
				}
				talentSqlAccessDetail = null;
				//Return results
				return affectedRows;
			}
			/// <summary>
			/// Deletes Configurations
			/// </summary>
			/// <param name="businessUnit"></param>
			/// <param name="paymentCode"></param>
			/// <param name="givenTransaction"></param>
			/// <returns></returns>
			/// <remarks></remarks>
			private int DeleteData(string businessUnit, string paymentCode, SqlTransaction givenTransaction)
			{
				int affectedRows = 0;
				TalentDataAccess talentSqlAccessDetail = new TalentDataAccess();
				ErrorObj err = new ErrorObj();
				string commandText = "DELETE FROM tbl_payment_type_bu WHERE business_unit=@BUSINESS_UNIT AND payment_type_code = @PAYMENT_TYPE_CODE";
				talentSqlAccessDetail.Settings = settings;
				talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteNonQuery;
				talentSqlAccessDetail.CommandElements.CommandText = commandText;
				talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BUSINESS_UNIT", businessUnit));
				talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@PAYMENT_TYPE_CODE", paymentCode));
				if (givenTransaction == null)
				{
					err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005);
				}
				else
				{
					err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005, givenTransaction);
				}
				if ((!(err.HasError)) && (!(talentSqlAccessDetail.ResultDataSet == null)))
				{
					affectedRows = System.Convert.ToInt32(talentSqlAccessDetail.ResultDataSet.Tables[0].Rows[0][0]);
				}
				talentSqlAccessDetail = null;
				return affectedRows;
			}
			/// <summary>
			/// Insert Configurations
			/// </summary>
			/// <param name="businessUnit"></param>
			/// <param name="paymentCode"></param>
			/// <param name="configs"></param>
			/// <param name="givenTransaction"></param>
			/// <returns></returns>
			/// <remarks></remarks>
			private int InsertData(string businessUnit, string paymentCode, System.Collections.Generic.List<ConfigurationItem> configs, SqlTransaction givenTransaction)
			{
				Dictionary<string, string> setValue = new Dictionary<string, string>();
				for (int i = 0; i <= configs.Count - 1; i++)
				{
					string name = configs[i].DefaultName;
					setValue.Add(name, string.Format("@{0}", name));
				}
				string columns = string.Format("({0})", string.Join(", ", setValue.Keys));
				string values = string.Format("VALUES({0})", string.Join(", ", setValue.Values));
				var commandText = string.Format("INSERT INTO tbl_payment_type_bu {0} {1}", columns, values);
				int affectedRows = 0;
				TalentDataAccess talentSqlAccessDetail = new TalentDataAccess();
				ErrorObj err = new ErrorObj();
				//Construct The Call
				talentSqlAccessDetail.Settings = settings;
				talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteNonQuery;
				talentSqlAccessDetail.CommandElements.CommandText = commandText;
				for (int i = 0; i <= configs.Count - 1; i++)
				{
					string paramName = string.Format("@{0}", configs[i].DefaultName);
					talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter(paramName, configs[i].UpdatedValue));
				}
				//Execute
				if (givenTransaction == null)
				{
					err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005);
				}
				else
				{
					err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005, givenTransaction);
				}
				if ((!(err.HasError)) && (!(talentSqlAccessDetail.ResultDataSet == null)))
				{
					affectedRows = System.Convert.ToInt32(talentSqlAccessDetail.ResultDataSet.Tables[0].Rows[0][0]);
				}
				talentSqlAccessDetail = null;
				//Return results
				return affectedRows;
			}
			#endregion
		}
	}
}

File: TalentCommon/DataObjects/Payment/tbl_payment_type_bu.vb
Imports System.Data.SqlClient
Namespace DataObjects.TableObjects
    ''' <summary>
    ''' Provides the functionality to manage data from the table tbl_payment_type_bu based on business functionality
    ''' </summary>
    <Serializable()> _
        Public Class tbl_payment_type_bu
        Inherits DBObjectBase

#Region "Class Level Fields"

        ''' <summary>
        ''' Instance of DESettings
        ''' </summary>
        Private _settings As New DESettings

        ''' <summary>
        ''' Class Name which is used in cache key construction
        ''' </summary>
        Const CACHEKEY_CLASSNAME As String = "tbl_payment_type_bu"
#End Region

#Region "Constructors"
        Sub New()
        End Sub
        ''' <summary>
        ''' Initializes a new instance of the <see cref="tbl_payment_type_bu" /> class.
        ''' </summary>
        ''' <param name="settings">The DESettings instance</param>
        Sub New(ByVal settings As DESettings)
            _settings = settings
        End Sub
#End Region

#Region "Private Methods"

#End Region

#Region "Public Methods"

        ''' <summary>
        ''' Gets all the business unit records from the table tbl_payment_type_bu
        ''' </summary>
        ''' <param name="cacheing">if set to <c>true</c> [cacheing].</param>
        ''' <param name="cacheTimeMinutes">The cache time minutes.</param>
        ''' <returns>DataSet with records</returns>
        Public Function GetAll(Optional ByVal cacheing As Boolean = True, Optional ByVal cacheTimeMinutes As Integer = 30) As DataTable

            Dim outputDataTable As New DataTable
            Dim cacheKeyPrefix As String = GetCacheKeyPrefix(CACHEKEY_CLASSNAME, "GetAll")
            Dim talentSqlAccessDetail As New TalentDataAccess
            Try
                'Construct The Call
                talentSqlAccessDetail.Settings = _settings
                talentSqlAccessDetail.Settings.Cacheing = cacheing
                talentSqlAccessDetail.Settings.CacheTimeMinutes = cacheTimeMinutes
                talentSqlAccessDetail.Settings.CacheStringExtension = cacheKeyPrefix
                talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteDataSet
                talentSqlAccessDetail.CommandElements.CommandText = "Select * From TBL_PAYMENT_TYPE_BU"

                'Execute
                Dim err As New ErrorObj
                err = talentSqlAccessDetail.SQLAccess()
                If (Not (err.HasError)) And (Not (talentSqlAccessDetail.ResultDataSet Is Nothing)) Then
                    outputDataTable = talentSqlAccessDetail.ResultDataSet.Tables(0)
                End If
            Catch ex As Exception
                Throw
            Finally
                talentSqlAccessDetail = Nothing
            End Try

            'Return the results
            Return outputDataTable

        End Function

        ''' <summary>
        ''' Gets the business unit records based on the given business unit.
        ''' </summary>
        ''' <param name="businessUnit">The business unit.</param>
        ''' <param name="cacheing">if set to <c>true</c> [cacheing]. Default <c>true</c></param>
        ''' <param name="cacheTimeMinutes">The cache time minutes. Default <c>30</c></param>
        ''' <returns>DataTable</returns>
        Public Function GetByBU(ByVal businessUnit As String, Optional ByVal cacheing As Boolean = True, Optional ByVal cacheTimeMinutes As Integer = 30) As DataTable

            Dim outputDataTable As New DataTable
            Dim cacheKeyPrefix As String = GetCacheKeyPrefix(CACHEKEY_CLASSNAME, "GetByBU")
            Dim talentSqlAccessDetail As New TalentDataAccess

            Try
                'Construct The Call
                talentSqlAccessDetail.Settings = _settings
                talentSqlAccessDetail.Settings.Cacheing = cacheing
                talentSqlAccessDetail.Settings.CacheTimeMinutes = cacheTimeMinutes
                talentSqlAccessDetail.Settings.CacheStringExtension = cacheKeyPrefix & ToUpper(businessUnit)
                talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteDataSet
                talentSqlAccessDetail.CommandElements.CommandText = "Select * From TBL_PAYMENT_TYPE_BU Where BUSINESS_UNIT=@BusinessUnitName"
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BusinessUnitName", businessUnit))

                'Execute
                Dim err As New ErrorObj
                err = talentSqlAccessDetail.SQLAccess()
                If (Not (err.HasError)) And (Not (talentSqlAccessDetail.ResultDataSet Is Nothing)) Then
                    outputDataTable = talentSqlAccessDetail.ResultDataSet.Tables(0)
                End If
            Catch ex As Exception
                Throw
            Finally
                talentSqlAccessDetail = Nothing
            End Try

            'Return the results 
            Return outputDataTable

        End Function


        ''' <summary>
        ''' Gets the business unit records based on the given business unit and the type of basket (retail/ticketing/mixed) and Agent mode.
        ''' Selects from tbl_payment_type_bu and tbl_payment_type
        ''' </summary>
        ''' <param name="businessUnit">The business unit.</param>
        ''' <param name="isMixedType">The boolean value to specify mixed type</param>
        ''' <param name="isRetailType">The boolean value to specify retail type</param>
        ''' <param name="isTicketingType">The boolean value to specify ticketing type</param>
        ''' <param name="isAgent">The boolean value to specify if the current user is in agent mode</param>
        ''' <param name="cacheing">if set to <c>true</c> [cacheing]. Default <c>true</c></param>
        ''' <param name="cacheTimeMinutes">The cache time minutes. Default <c>30</c></param>
        ''' <returns>DataTable</returns>
        Public Function GetByBasketTypeAndBU(ByVal businessUnit As String, ByVal isTicketingType As Boolean, ByVal isRetailType As Boolean, ByVal isMixedType As Boolean, _
                                             ByVal isAgent As Boolean, ByVal isRefundMode As Boolean, ByVal isGenericSales As Boolean,
                                             Optional ByVal cacheing As Boolean = True, Optional ByVal cacheTimeMinutes As Integer = 30,
                                                Optional ByVal isOtherType As Boolean = False) As DataTable

            Dim outputDataTable As New DataTable
            Dim cacheKeyPrefix As String = GetCacheKeyPrefix(CACHEKEY_CLASSNAME, "GetByBasketTypeAndBU")
            Dim talentSqlAccessDetail As New TalentDataAccess
            Dim commandText As System.Text.StringBuilder = New System.Text.StringBuilder

            Try
                'Construct The Call
                talentSqlAccessDetail.Settings = _settings
                talentSqlAccessDetail.Settings.Cacheing = cacheing
                talentSqlAccessDetail.Settings.CacheTimeMinutes = cacheTimeMinutes
                talentSqlAccessDetail.Settings.CacheStringExtension = cacheKeyPrefix & ToUpper(businessUnit) & ToUpper(isTicketingType.ToString) & _
                                                                    ToUpper(isRetailType.ToString) & ToUpper(isMixedType.ToString) & ToUpper(isAgent.ToString) & _
                                                                    ToUpper(isRefundMode.ToString) & ToUpper(isGenericSales.ToString) & ToUpper(isOtherType.ToString)
                talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteDataSet
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BusinessUnitName", businessUnit))
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@IsOtherType", isOtherType, SqlDbType.Bit))
                commandText.Append(" SELECT PTBU.QUALIFIER, PTBU.PARTNER, PTBU.PAYMENT_TYPE_CODE, PTBU.DEFAULT_PAYMENT_TYPE, ")
                commandText.Append(" PT.PAYMENT_TYPE_DESCRIPTION, PT.SEQUENCE, PT.NAVIGATE_URL, PT.CALL_CREDIT_LIMIT_CHECK, PPS_TYPE, IS_GIFT_CARD ")
                commandText.Append(" FROM tbl_payment_type_bu PTBU, tbl_payment_type PT ")
                commandText.Append(" WHERE PTBU.BUSINESS_UNIT = @BusinessUnitName ")
                commandText.Append(" AND PT.IS_OTHER_TYPE = @IsOtherType ")
                commandText.Append(" AND PTBU.PAYMENT_TYPE_CODE = PT.PAYMENT_TYPE_CODE")
                If isAgent Then
                    If isRefundMode AndAlso isGenericSales Then
                        commandText.Append(" AND PTBU.GENERIC_SALES_REFUND = 1")
                    ElseIf isRefundMode Then
                        commandText.Append(" AND PTBU.AGENT_TICKETING_TYPE_REFUND = 1")
                    ElseIf isGenericSales Then
                        commandText.Append(" AND PTBU.GENERIC_SALES = 1")
                    Else
                        If isMixedType Then
                            commandText.Append(" AND PTBU.AGENT_RETAIL_TYPE = 1 AND PTBU.AGENT_TICKETING_TYPE = 1")
                        ElseIf isRetailType Then
                            commandText.Append(" AND PTBU.AGENT_RETAIL_TYPE = 1")
                        ElseIf isTicketingType Then
                            commandText.Append(" AND PTBU.AGENT_TICKETING_TYPE = 1")
                        End If
                    End If
                Else
                    If isRefundMode Then
                        commandText.Append(" AND PTBU.TICKETING_TYPE_REFUND = 1")
                    Else
                        If isMixedType Then
                            commandText.Append(" AND PTBU.RETAIL_TYPE = 1 AND PTBU.TICKETING_TYPE = 1")
                        ElseIf isRetailType Then
                            commandText.Append(" AND PTBU.RETAIL_TYPE = 1")
                        ElseIf isTicketingType Then
                            commandText.Append(" AND PTBU.TICKETING_TYPE = 1")
                        End If
                    End If

                End If
                talentSqlAccessDetail.CommandElements.CommandText = commandText.Append(" ORDER BY PT.SEQUENCE").ToString()

                'Execute
                Dim err As New ErrorObj
                err = talentSqlAccessDetail.SQLAccess()
                If (Not (err.HasError)) And (Not (talentSqlAccessDetail.ResultDataSet Is Nothing)) Then
                    outputDataTable = talentSqlAccessDetail.ResultDataSet.Tables(0)
                End If
            Catch ex As Exception
                Throw
            Finally
                talentSqlAccessDetail = Nothing
            End Try

            'Return the results 
            Return outputDataTable

        End Function

        ''' <summary>
        ''' Deletes the specified business unit.
        ''' </summary>
        ''' <param name="businessUnitToDelete">The business unit to delete.</param>
        ''' <param name="givenTransaction">The given transaction.</param>
        ''' <returns>No of affected rows</returns>
        Public Function Delete(ByVal businessUnitToDelete As String, Optional ByVal givenTransaction As SqlTransaction = Nothing) As Integer

            Dim affectedRows As Integer = 0
            Dim talentSqlAccessDetail As New TalentDataAccess
            Dim err As New ErrorObj
            'Construct The Call
            talentSqlAccessDetail.Settings = _settings
            talentSqlAccessDetail.Settings.Cacheing = False
            talentSqlAccessDetail.Settings.CacheStringExtension = ""
            talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteNonQuery
            Dim sqlStatement As String = "DELETE TBL_PAYMENT_TYPE_BU " & _
                "WHERE BUSINESS_UNIT=@BusinessUnitToDelete "
            talentSqlAccessDetail.CommandElements.CommandText = sqlStatement
            talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BusinessUnitToDelete", businessUnitToDelete))

            'Execute
            If (givenTransaction Is Nothing) Then
                err = talentSqlAccessDetail.SQLAccess()
            Else
                err = talentSqlAccessDetail.SQLAccess(givenTransaction)
            End If
            If (Not (err.HasError)) And (Not (talentSqlAccessDetail.ResultDataSet Is Nothing)) Then
                affectedRows = talentSqlAccessDetail.ResultDataSet.Tables(0).Rows(0)(0)
            End If
            talentSqlAccessDetail = Nothing

            'Return results
            Return affectedRows

        End Function

        ''' <summary>
        ''' Copies the records from one business unit to another business unit inside the table tbl_payment_type_bu
        ''' irrespective of whether destination business exists or not and returns no of affected rows
        ''' </summary>
        ''' <param name="fromBusinessUnit">From business unit.</param>
        ''' <param name="toBusinessUnit">To business unit.</param>
        ''' <returns>No Of Affected Rows</returns>
        Public Function CopyByBU(ByVal fromBusinessUnit As String, ByVal toBusinessUnit As String, Optional ByVal givenTransaction As SqlTransaction = Nothing) As Integer

            Dim affectedRows As Integer = 0
            Dim talentSqlAccessDetail As New TalentDataAccess


            'Construct The Call
            talentSqlAccessDetail.Settings = _settings
            talentSqlAccessDetail.Settings.Cacheing = False
            talentSqlAccessDetail.Settings.CacheStringExtension = ""
            talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteNonQuery
            Dim sqlStatement As String = "INSERT INTO TBL_PAYMENT_TYPE_BU (BUSINESS_UNIT, QUALIFIER, PARTNER, " & _
                        "PAYMENT_TYPE_CODE, DEFAULT_PAYMENT_TYPE) " & _
                        "SELECT @ToBusinessUnit As BUSINESS_UNIT, QUALIFIER, PARTNER, " & _
                        "PAYMENT_TYPE_CODE, DEFAULT_PAYMENT_TYPE FROM TBL_PAYMENT_TYPE_BU " & _
                        "WHERE BUSINESS_UNIT = @FromBusinessUnit"

            talentSqlAccessDetail.CommandElements.CommandText = sqlStatement
            talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@FromBusinessUnit", fromBusinessUnit))
            talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@ToBusinessUnit", toBusinessUnit))

            'Execute
            Dim err As New ErrorObj
            If (givenTransaction Is Nothing) Then
                err = talentSqlAccessDetail.SQLAccess()
            Else
                err = talentSqlAccessDetail.SQLAccess(givenTransaction)
            End If
            If (Not (err.HasError)) And (Not (talentSqlAccessDetail.ResultDataSet Is Nothing)) Then
                affectedRows = talentSqlAccessDetail.ResultDataSet.Tables(0).Rows(0)(0)
            End If

            'Return the results 
            Return affectedRows

        End Function

#End Region

    End Class
End Namespace